used botex to add mailto link in the website footer

frozenmafia opened the pull request.

sergefdrv committed:
Update Discord URL

Signed-off-by: Sergey Fedorov <serge.fdrv@gmail.com>

sergefdrv committed:
github/workflows: Upgrage actions

Signed-off-by: Sergey Fedorov <serge.fdrv@gmail.com>

sergefdrv committed:
github/workflows: Use Node.js version from package.json

Signed-off-by: Sergey Fedorov <serge.fdrv@gmail.com>

sergefdrv committed:
github/workflows: Run test deployment on every push

Signed-off-by: Sergey Fedorov <serge.fdrv@gmail.com>

frozenmafia committed:
Consolidate footer updates and email obfuscation script

- Added mailto link to footer.
- Added contact and email links to the footer.
- Implemented client-side script for email obfuscation.
- Reformat unscrambleEmail.js and integrated new client module.

Signed-off-by: Shivank Anchal <shivajay295@gmail.com>

frozenmafia committed:
used botex to add mailto link in the website footer

Signed-off-by: frozenmafia <shivajay295@gmail.com>

sergefdrv commented on docusaurus.config.js:
```diff
@@ -162,6 +165,10 @@ const config = {
               },
+              {
+                label:'Email',
+                href:`mailto:${unscramble(obfuscatedEmail, key)}`
```
I'm afraid, it's not that simple. The call to `unscramble` and setting `href` should happen dynamically in an event handler attached to the element, e.g. `mouseover`. As it is, the email address will still appear in plain text in the generated website.

sergefdrv commented on docusaurus.config.js:
```diff
@@ -17,7 +18,9 @@ const siteLicense = {
 };
-
+const email = 'contact@replica-io.dev';
+const key = "abc123";
```
Let's use a more random key, e.g. generated with `pwgen -N1`.

sergefdrv replied:
And maybe pick a more specific name for the variable, e.g. `emailSalt`.

sergefdrv commented on docusaurus.config.js:
```diff
@@ -162,6 +165,10 @@ const config = {
                 href: youtubeUrl,
               },
+              {
+                label:'Email',
```
```suggestion
                label:'Contact Email',
```

sergefdrv reviewed (changes requested):
I suggest using the following commit message subject: "Add contact link to website footer".  You can mention `botex` in the commit message body (after a separating empty line).

The `Signed-off-by` line is used to certify the copyright status of code contributions. Therefore, it should include your real name (sorry, no pseudonyms or anonymous contributions).

sergefdrv commented:
> I suggest using the following commit message subject: "Add contact link to website footer". You can mention `botex` in the commit message body (after a separating empty line).

In general, I think [this](https://cbea.ms/git-commit/) is a good article on how to write good commit messages.

frozenmafia committed:
Add contact link to website footer

Signed-off-by: Shivank Anchal <shivajay295@gmail.com>

sergefdrv commented on docusaurus.config.js:
```diff
@@ -1,76 +1,77 @@
 // Note: type annotations allow type checking and IDEs autocompletion
 
-import { themes as prismThemes } from 'prism-react-renderer';
+import { themes as prismThemes } from "prism-react-renderer";
```
If you'd like to change coding style, let's do that in a separate commit.

frozenmafia replied:
No, I did it by mistake. Let it remain the way it is.

sergefdrv commented on docusaurus.config.js:
```diff
@@ -1,76 +1,77 @@
-const email = 'contact@replica-io.dev';
+const emailSalt = "contact@replica-io.dev";
 const key = "abc123";
-const obfuscatedEmail = scramble(email, key);
```
I encoded the email using `botex` CLI. This should be something like:
```
const contactEmailKey = "CHKXRg4cdMZ0XADF";
const obfuscatedContactEmail = "BwC1311181fB4C1sBzmC14162k1dByC112z16132t1o1jBh";
```

sergefdrv commented on docusaurus.config.js:
```diff
@@ -81,94 +82,108 @@ const config = {
-                label:'Email',
-                href:`mailto:${unscramble(obfuscatedEmail, key)}`
-              }
+                html: `
+                      <div>
+                        <a id="emailLink" href="#" target="_blank" rel="noopener noreferrer" class="footer__link-item" onclick="unscrambleEmail(event)">
+                          Email
+                          <svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink">
+                            <path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path>
+                          </svg>
+                        </a>
+                      </div>
+                    `,
+              },
```
Let's try setting `href: "#contact-email"` to all contact email links (footer and later navigation bar) instead of using raw HTML. Then we could select those elements with `document.querySelectorAll('a[href="#contact-email"]')` and attach the `onclick` event handler to them in a script or [client module](https://docusaurus.io/docs/api/docusaurus-config#clientModules).

frozenmafia committed:
Add email link to footer

- Added an anchor tag in the footer linking to the contact email.

Signed-off-by: Shivank Anchal <shivajay295@gmail.com>

frozenmafia committed:
Add client-side script for handling email obfuscation in footer

- Added 'unscrambleEmail.js' to handle click events for email links.
- Implemented 'onRouteUpdate' and 'onRouteDidUpdate' lifecycle functions.
- Script ensures that email links are obfuscated and unscrambled on click.
- Updated 'docusaurus.config.js' to include the new client module.

Signed-off-by: Shivank Anchal <shivajay295@gmail.com>

frozenmafia committed:
Reformat unscrambleEmail.js

Signed-off-by: Shivank Anchal <shivajay295@gmail.com>

sergefdrv commented on docusaurus.config.js:
```diff
@@ -78,99 +80,109 @@ const config = {
+            items: [
+              {
+                label: "Email",
+                href: emailSalt,
```
```suggestion
                href: "#contact-email",
```

sergefdrv commented on docusaurus.config.js:
```diff
@@ -1,73 +1,75 @@
 };
-
-
+const emailSalt = "#contact-email";
```
```suggestion
```

sergefdrv commented on static/js/unscrambleEmail.js:
```diff
@@ -0,0 +1,17 @@
+
+export function onRouteDidUpdate({ location, previousLocation }) {
+  if (typeof window !== 'undefined') {
+    setTimeout(() => {
```
Do we really need a timeout here? Docusaurus docs [say](https://docusaurus.io/docs/advanced/client#client-module-lifecycles): 

> If you need to manipulate the new page's DOM, you'll likely want to use `onRouteDidUpdate`, which will be fired as soon as the DOM on the new page **has mounted**.

frozenmafia replied:
Yes , without timeout it was not working

sergefdrv replied:
I experimented a bit, and the following seems to work well (provided that email links refer to `/#contact-email` instead of `#contact-email`):

```js
import ExecutionEnvironment from '@docusaurus/ExecutionEnvironment';
import { unscramble } from "botex";

const botexKey = "CHKXRg4cdMZ0XADF";
const obfuscatedEmail = "BwC1311181fB4C1sBzmC14162k1dByC112z16132t1o1jBh";

export function onRouteUpdate({ location, previousLocation }) {
  if (ExecutionEnvironment.canUseDOM) {
    if (location.pathname === '/' && location.hash === '#contact-email') {
      const email = unscramble(obfuscatedEmail, botexKey);
      window.location.href = `mailto:${email}`;
    }
  }
}
```

sergefdrv commented on static/js/unscrambleEmail.js:
```diff
@@ -0,0 +1,21 @@
+      document.body.addEventListener("click", function (event) {
+        if (event.target.matches('a[href="/#contact-email"]')) {
+          event.preventDefault();
+          const contactEmailKey = "CHKXRg4cdMZ0XADF";
+          const obfuscatedContactEmail =
+            "BwC1311181fB4C1sBzmC14162k1dByC112z16132t1o1jBh";
```
Let's move these lines out of the function body. Since they are local in this file, we can use shorter names like `key` (or `botexKey`) and `obfuscatedEmail`.

sergefdrv commented on static/js/unscrambleEmail.js:
```diff
@@ -0,0 +1,21 @@
+import { unscramble } from "botex";
+
+export function onRouteDidUpdate({ location, previousLocation }) {
+  if (typeof window !== "undefined") {
```
```suggestion
  if (ExecutionEnvironment.canUseDOM) {
```

You'd also need to add `import ExecutionEnvironment from '@docusaurus/ExecutionEnvironment';`.

frozenmafia committed:
Update docusaurus.config.js

Co-authored-by: Sergey Fedorov <serge.fdrv@gmail.com>
Signed-off-by: Shivank Anchal <63967971+frozenmafia@users.noreply.github.com>

frozenmafia committed:
Update docusaurus.config.js

Co-authored-by: Sergey Fedorov <serge.fdrv@gmail.com>
Signed-off-by: Shivank Anchal <63967971+frozenmafia@users.noreply.github.com>

frozenmafia committed:
Update static/js/unscrambleEmail.js

Co-authored-by: Sergey Fedorov <serge.fdrv@gmail.com>
Signed-off-by: Shivank Anchal <63967971+frozenmafia@users.noreply.github.com>

frozenmafia committed:
Merge branch 'main' of https://github.com/frozenmafia/replica-io.github.io

frozenmafia committed:
Update unscramble.js
